README.md
# shared-canvas

An application that allows multiple users to draw on the same canvas.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-10-24
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-10-24
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/shared-canvas-server/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/shared-canvas-server/compare/0.1.0...0.1.1


test/shared_canvas/core_test.clj
(ns shared-canvas.core-test
  (:require [clojure.test :refer :all]
            [shared-canvas.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


test/shared_canvas_server/core_test.clj
(ns shared-canvas-server.core-test
  (:require [clojure.test :refer :all]
            [shared-canvas-server.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/shared_canvas/core.clj
(ns shared-canvas.core
  (:gen-class)
  (:require [org.httpkit.server :refer :all]
            [compojure.core :refer [GET defroutes]]
            [compojure.route :refer [not-found]]
            [clojure.data.json :as json]
            [ring.middleware.resource :refer [wrap-resource]]
            [ring.middleware.content-type :refer [wrap-content-type]]
            [ring.middleware.not-modified :refer [wrap-not-modified]]
            [environ.core :refer [env]])) 

;;;; Configuration

(defonce port (-> (or (env :port) "8000") Integer/parseInt))

;;;; State

(defonce canvas-events (atom []))
(defonce channels (atom {}))

;;;; Handlers

(defn handle-message
  [message]
  (println "Received:" message)
  (swap! canvas-events conj (json/read-str message))
  (doseq [channel (keys @channels)]
    (send! channel message)))

(defn websocket-handler
  [request]
  (with-channel request channel
    (println "New user connected")
    (swap! channels assoc channel request)
    (println "Sending canvas events")
    (send! channel (json/write-str @canvas-events))
    (on-close channel (fn [status] (swap! channels dissoc channel)))
    (on-receive channel handle-message)))

;;;; Middleware

(defn wrap-root-to-index
  [handler]
  (fn [request]
    (handler (update-in request [:uri]
                #(if (= "/" %) "/index.html" %)))))

;;;; Routes

(defroutes routes
  (GET "/ws" [] websocket-handler)
  (not-found "Not found"))

(def app
  (-> routes
      (wrap-resource "public")
      (wrap-content-type)
      (wrap-not-modified)
      (wrap-root-to-index)))

;;;; Server

(defonce server (atom nil))

(defn stop-server []
  (when-not (nil? @server)
    (@server :timeout 100)
    (reset! server nil)))

(defn start-server
  [port]
  (reset! server (run-server #'app {:port port})))

(defn -main
  [& args]
  (start-server port)
  (println "Server running on port 8000"))


resources/public/banned-drawings.js
(function () {

  //// Websocket functions

  function onOpen(event) {
    console.log(`Websocket channel opened on ${event.target.url}`, event);
  }

  function onClose(event) {
    console.log('Websocket channel closed:', event);
  }

  function onMessage(event) {
    console.log('Received message:', JSON.parse(event.data));
  }

  function onError(event) {
    console.log('Error:', event);
  }

  const websocket = new WebSocket("ws://localhost:8000/ws");
  websocket.onopen = onOpen;
  websocket.onclose = onClose;
  websocket.onmessage = onMessage;
  websocket.onerror = onError;

  //// Get the canvas element and context

  const canvas = document.getElementById('board');
  const canvasContext = canvas.getContext('2d');
  
  //// Initialize context

  canvasContext.fillStyle = 'rgba(0, 0, 200, 0.5)';

  //// Configuration

  const clickEvent = null;
  const mouseMoveEvent = null;
  const drawWidth = 50;
  const drawHeight = 50;
  var mouseDown = false;

  //// Utility functions

  function getMousePosition(canvas, event) {
    const rect = canvas.getBoundingClientRect();
    return {
      x: event.clientX - rect.left,
      y: event.clientY - rect.top
    };
  }

  function drawRectangle(ctx, mousePosition, width, height) {
    console.log('drawing rectangle');
    const x = mousePosition.x - (width / 2);
    const y = mousePosition.y - (height / 2);
    ctx.fillRect(x, y, width, height);
    websocket.send(JSON.stringify({
      x,
      y,
      width,
      height,
      fillStyle: ctx.fillStyle
    }));
  }

  //// Event listeners

  canvas.addEventListener('click', function (event) {
    drawRectangle(canvasContext, getMousePosition(canvas, event), drawWidth, drawHeight);
  });

  canvas.addEventListener('mousedown', function (event) {
    mouseDown = true;
  });

  canvas.addEventListener('mouseup', function (event) {
    mouseDown = false;
  });

  canvas.addEventListener('mousemove', function(event) {
    if (mouseDown) {
      drawRectangle(canvasContext, getMousePosition(canvas, event), drawWidth, drawHeight);
    }
  });  
})();




resources/public/canvas.js
(function () {

  //// Websocket functions

  function onOpen(event) {
    console.log(`Websocket channel opened on ${event.target.url}`, event);
  }

  function onClose(event) {
    console.log('Websocket channel closed:', event);
  }

  function onMessage(event) {
    const message = JSON.parse(event.data);
    console.log('Received message:', message);
    if (message.constructor === Array) {
      message.forEach(drawEvent => {
        drawRectangle(canvasContext, drawEvent);
      });
    } else if (message.constructor === Object) {
      drawRectangle(canvasContext, message);
    }
  }

  function onError(event) {
    console.log('Error:', event);
  }

  const websocket = new WebSocket(`ws://${window.location.host}/ws`);
  websocket.onopen = onOpen;
  websocket.onclose = onClose;
  websocket.onmessage = onMessage;
  websocket.onerror = onError;

  //// Get the canvas element and context

  const canvas = document.getElementById('canvas');
  const canvasContext = canvas.getContext('2d');

  //// State

  let mouseDown = false;
  let drawWidth = 50;
  let drawHeight = 50;
  let fillStyle = 'rgba(0, 0, 200, 0.5)';

  //// Initialize context

  canvasContext.fillStyle = fillStyle;

  //// Utility functions

  function getMousePosition(canvas, event) {
    const rect = canvas.getBoundingClientRect();
    return {
      x: event.clientX - rect.left,
      y: event.clientY - rect.top
    };
  }

  function drawRectangle(ctx, drawEvent) {
    console.log('drawing rectangle');
    ctx.fillStyle = drawEvent.fillStyle;
    ctx.fillRect(drawEvent.x, drawEvent.y, drawEvent.width, drawEvent.height);
  }

  function createDrawEvent(mousePosition) {
    return {
      x: mousePosition.x - (drawWidth / 2),
      y: mousePosition.y - (drawHeight / 2),
      width: drawWidth,
      height: drawHeight,
      fillStyle
    };
  }

  //// Event listeners

  canvas.addEventListener('click', function (event) {
    const drawEvent = createDrawEvent(getMousePosition(canvas, event));
    drawRectangle(canvasContext, drawEvent);
    websocket.send(JSON.stringify(drawEvent));
  });

  canvas.addEventListener('mousedown', function (event) {
    mouseDown = true;
  });

  canvas.addEventListener('mouseup', function (event) {
    mouseDown = false;
  });

  canvas.addEventListener('mousemove', function (event) {
    if (mouseDown) {
      const drawEvent = createDrawEvent(getMousePosition(canvas, event));
      drawRectangle(canvasContext, drawEvent);
      websocket.send(JSON.stringify(drawEvent));
    }
  });
})();


